### PR TITLE
Sequence annotation

### PR DESCRIPTION
**Pull request by macieksmuga**
Implementation of Sequence Annotation API, including `FeatureSet` and `Feature` records as well as get & search endpoints for each. Corresponds to https://github.com/ga4gh/schemas/pull/591 and https://github.com/ga4gh/compliance/pull/181

### CODE REVIEW

**Comment by macieksmuga** [comment]
- [ ] Note to self: I had added a new integration test for the gff3 ETL script, somehow didn't make it into this PR. Fix this.

**Comment by david4096** [comment]
I believe this is a problem with the request validation, in case someone else runs into it:

```
HTTP/1.0 400 BAD REQUEST
Content-Length: 285
Content-Type: application/json
Date: Fri, 01 Apr 2016 21:23:24 GMT
Server: Werkzeug/0.11.5 Python/2.7.11

{
    "errorCode": 2023404236, 
    "message": "Request '{u'start': 0, u'parentId': u'', u'end': 10000000000, u'featureSetId': u'YnJjYTE6Z2VuY29kZXYxOQ', u'referenceName': u'ref_brca1'}' is not a valid instance of <class 'ga4gh._protocol_definitions.SearchFeaturesRequest'>; invalid fields: {}"
}
```

`featureTypes` is a required field (that can be left empty) according to the schemas but it isn't reported here.

However, parentId is nullable according to the schemas and causes validation to fail:

```
HTTP/1.0 400 BAD REQUEST
Content-Length: 296
Content-Type: application/json
Date: Fri, 01 Apr 2016 21:25:32 GMT
Server: Werkzeug/0.11.5 Python/2.7.11

{
    "errorCode": 2023404236, 
    "message": "Request '{u'start': 0, u'end': 10000000000, u'featureSetId': u'YnJjYTE6Z2VuY29kZXYxOQ', u'referenceName': u'ref_brca1'}' is not a valid instance of <class 'ga4gh._protocol_definitions.SearchFeaturesRequest'>; invalid fields: {u'parentId': u'MISSING_VALUE'}"
}
```

**Comment by david4096** [comment]
The logging is getting set incorrectly because of `sqliteBackend.py`.

**Comment by macieksmuga** [comment]
@david4096 Thanks for running the code!
Some answers:

- [ ] incorrect logging is already being addressed in #1048 
- [ ] Good catch on the missing `featureTypes`. The server needs to throw a validation error if this is not provided. Will fix.

- [ ] ditto on `parentId`. Will fix.

- [ ] the schemas are silent on start > end, so behavior is undefined in this case. I _should_ be throwing an error in that case, but (alas) am not breaking schemas as is. Will fix.

**Comment by diekhans** [comment]
added `read/variant overlap search does not specify behavior of
aberrant coordinates.'  https://github.com/ga4gh/schemas/issues/597

to resolve overlap search behavior issue.

**Comment by macieksmuga** [comment]
Updating my comment  - It's weirder than I thought. I'm deploying the server on the compliance dataset (per https://github.com/ga4gh/compliance/pull/181) and here's what's happening:

- on request `curl -XPOST -H "Content-Type:application/json" -d'{"featureSetId": "YnJjYTE6Z2VuY29kZXYxOQ", "referenceName": "ref_brca1", "start": 0, "end": 10, "featureTypes": []}' localhost:8000/features/search`, I get a valid response.

- on request `curl -XPOST -H "Content-Type:application/json" -d'{"featureSetId": "YnJjYTE6Z2VuY29kZXYxOQ", "referenceName": "ref_brca1", "start": 0, "end": 10}' localhost:8000/features/search` I get the error: 
```
RequestValidationFailureException: Request '{u'start': 0, u'end': 10, u'featureSetId': u'YnJjYTE6Z2VuY29kZXYxOQ', u'referenceName': u'ref_brca1'}' is not a valid instance of <class 'ga4gh._protocol_definitions.SearchFeaturesRequest'>; invalid fields: {u'parentId': u'MISSING_VALUE'}
```
(it picks the wrong field to complain about)

- on request `curl -XPOST -H "Content-Type:application/json" -d'{"featureSetId": "YnJjYTE6Z2VuY29kZXYxOQ", "referenceName": "ref_brca1", "start": 0, "end": 10, "parentId": "YnJjYTE6Z2VuY29kZXYxOTo0MzQ3MjA2Njcy"}' localhost:8000/features/search`, I get the error:
```
RequestValidationFailureException: Request '{u'start': 0, u'parentId': u'YnJjYTE6Z2VuY29kZXYxOTo0MzQ3MjA2Njcy', u'end': 10, u'featureSetId': u'YnJjYTE6Z2VuY29kZXYxOQ', u'referenceName': u'ref_brca1'}' is not a valid instance of <class 'ga4gh._protocol_definitions.SearchFeaturesRequest'>; invalid fields: {}
```
(which tells me nothing).

- on request `curl -XPOST -H "Content-Type:application/json" -d'{"referenceName": "ref_brca1", "start": 0, "end": 10, "parentId": "YnJjYTE6Z2VuY29kZXYxOTo0MzQ3MjA2Njcy"}' localhost:8000/features/search`, I get the error:
```
RequestValidationFailureException: Request '{u'start': 0, u'parentId': u'YnJjYTE6Z2VuY29kZXYxOTo0MzQ3MjA2Njcy', u'end': 10, u'referenceName': u'ref_brca1'}' is not a valid instance of <class 'ga4gh._protocol_definitions.SearchFeaturesRequest'>; invalid fields: {u'featureSetId': u'MISSING_VALUE'}
```
which is wrong, as `featureSetId` can be omitted when `parentId` is presented.

So yeah, it's a validation mess.

**Comment by jeromekelleher** [comment]
@dcolligan, any thoughts on what's happening here validation-wise?

**Comment by dcolligan** [comment]
Hard to tell just by eyeballing it.  I would need to look into it.

**Comment by david4096** [comment]
I think this is a known issue.

https://github.com/ga4gh/server/pull/952

**Comment by david4096** [comment]
I turned off input validation to see whether the SA API still behaved appropriately.

Failing to provide a featureSetId works as expected:
```
HTTP/1.0 400 BAD REQUEST
Content-Length: 87
Content-Type: application/json
Date: Mon, 04 Apr 2016 22:19:19 GMT
Server: Werkzeug/0.11.5 Python/2.7.11

{
    "errorCode": 30818580, 
    "message": "One of featureSetId or parentId must be supplied."
}
```

I found that failing to provide both start and end coordinates would cause a 500. However, providing coordinates without a referenceName will return no results.

Sending a numeric referenceName, when it expects a string, returns no results.

Providing a malformed parentId failed gracefully:

```
{
    "errorCode": 1304817242, 
    "message": "The identifier provided is invalid: '1' "
}
```

Sending in a malformed featureTypes field can cause the server to Error:

```
➜  compliance git:(sequence-annotation) ✗ http post localhost:8000/features/search featureSetId=YnJjYTE6Z2VuY29kZS52MjQuYW5ub3RhdGlvbg start:=0 end:=123123123 referenceName=chr1 featureTypes:=1         
HTTP/1.0 500 INTERNAL SERVER ERROR
Content-Length: 61
Content-Type: application/json
Date: Mon, 04 Apr 2016 22:25:20 GMT
Server: Werkzeug/0.11.5 Python/2.7.11

{
    "errorCode": 1277500761, 
    "message": "Internal Server Error"
}
```

So, in general, the internal checks for SA are good. Removing input validation makes the interface more usable: one does not have to specify unused fields: parentId, featureTypes. Since the featureTypes field is an array it can be cumbersome to design a request and is therefore prone to input errors. Although there are other combinations to be tried, I was hoping to show that without input validation turned on this code could still be used, and it appears so!

File: scripts/generate_gff3_db.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+                    feature.attributes.get("transcript_name", [None])[0],
+                    _db_serialize(feature.attributes))
+                dbcur.execute((
+                    "INSERT INTO FEATURE ("
+                    "id, parent_id, child_ids, reference_name, "
+                    "source, type, start, end, score, strand, "
+                    "name, gene_name, transcript_name, attributes) "
+                    "VALUES(?,?,?,?,?,?,?,?,?,?,?,?,?,?)"), values)
+                dbconn.commit()
+        dbcur.close()
```
Comment (david4096) [inline scripts/generate_gff3_db.py]: Make and verify indices in this script somewhere. Gencode.v.24 is about 2.8GB.

`create INDEX idx1 on feature(start, end, reference_name);`
`PRAGMA INDEX_LIST('feature');`

**Comment by macieksmuga** [comment]
The latest update against master didn't take. Let's see what happens if we close & reopen this PR.

**Comment by macieksmuga** [comment]
Now, re-open.

**Comment by macieksmuga** [comment]
@jeromekelleher @dcolligan I believe this is mergeable once https://github.com/ga4gh/schemas/pull/591 is merged.

**Comment by jeromekelleher** [comment]
Yep, +1 to merge. Great work on getting this in @macieksmuga and @david4096!

**Comment by jeromekelleher** [comment]
Can you rebase this against master and resolve the conflicts please?

**Pull request merged**